src/hs/Vehicle/Core/Normalise.hs

```diff
@@ -0,0 +1,109 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE DataKinds #-}
+
+module Vehicle.Core.Normalise
+  ( Norm
+  , NormError (..)
+  , MonadNorm
+  , runNorm
+  ) where
+
+import Vehicle.Core.Type ( Sort (..), Expr (..), Sort, EArg (..) )
+import Vehicle.Core.Instance.Recursive (ExprF (..))
+import Control.Monad.Except (MonadError, Except, runExcept)
+import Data.Map (Map, lookup, insert, notMember)
+import Data.Functor.Foldable (fold)
+import Vehicle.Core.Abs (ExprName, SortedBuiltin)
+import Control.Monad.Error.Class (throwError)
+
+-- * Normalisation monad
+
+type DeBruijn = Int
+
+newtype SortedDeBruijn (sort :: Sort) 
+  = SortedDeBruijn DeBruijn
+
+type NormExpr ann = Expr SortedDeBruijn SortedBuiltin ann
+
+-- |Errors thrown during normalisation
+data NormError = MissingDefFunType ExprName
+    | UnboundVariableError DeBruijn
+    | DuplicateVariableError DeBruijn
+    | MalformedLambdaError
+
+-- |Constraint for the monad stack used by the normaliser.
+type MonadNorm m = MonadError NormError m
+
+-- | Run a function in 'MonadNorm'.
+runNorm :: Except NormError a -> Either NormError a
+runNorm = runExcept
+
+-- |Class for the various normalisation functions.
+-- Invariant is that everything in the context is fully normalised
+class Norm vf where
+  norm :: MonadNorm m => Map DeBruijn vf -> vf -> m vf
+
+instance Norm (NormExpr ann) where
+  norm ctxt = fold $ \case
+    EVarF ann (SortedDeBruijn i) -> normVar ctxt ann i
+    ELetF _ arg exp1 exp2 -> normLet ctxt (index arg) exp1 exp2
+    EAppF _ exp1 exp2 -> normApp ctxt exp1 exp2
+
+    ELitSeqF ann exprs -> ELitSeq ann <$> sequence exprs
+    EAnnF ann expr typ -> EAnn ann <$> expr <*> pure typ
+    ELamF ann arg expr -> ELam ann arg <$> expr
+    ETyLamF ann targ expr -> ETyLam ann targ <$> expr
+    ETyAppF ann expr typ -> ETyApp ann <$> expr <*> pure typ    
+
+    ELitIntF ann val -> return $ ELitInt ann val
+    ELitRealF ann val -> return $ ELitReal ann val
+    EConF ann builtin -> return $ ECon ann builtin
+
+normApp :: (MonadNorm m) => Map DeBruijn (NormExpr ann) -> 
+            m (NormExpr ann) -> m (NormExpr ann) -> m (NormExpr ann) 
+normApp ctxt mFn mArg = do
+    fn <- mFn
+    arg <- mArg
+    varName <- lamVarName fn
+    varBody <- lamBody fn
+    updatedCtxt <- updateCtxt ctxt varName arg
+    norm updatedCtxt varBody 
```

Comment:
Bit worried about this call and the similar one in `normLet`. If I understand the semantics of the `Vehicle.Core.Instance.Recursive` the fold operator is already normalising all the sub-expressions for us in each case. That means that this is the second time the body is being normalised. The problem is of course that it's only in this case that we have the function's bound variable in the context....

Does that mean I can't use the recursive instance available in `Vehicle.Core.Instance.Recursive`?